libevm/VMOpt.cpp

```diff
@@ -49,175 +49,237 @@ void VM::initMetrics()
 	m_codeSpace.resize(extendedSize);
 	m_code = m_codeSpace.data();
 }
 
-// Implementation of EXP.
-//
-// This implements exponentiation by squaring algorithm.
-// Is is faster than boost::multiprecision::powm() because avoids explicit
-// mod operation.
-// Do not inline it.
-u256 VM::exp256(u256 _base, u256 _exponent)
-{
-	using boost::multiprecision::limb_type;
-	u256 result = 1;
-	while (_exponent)
-	{
-		if (static_cast<limb_type>(_exponent) & 1)  // If exponent is odd.
-			result *= _base;
-		_base *= _base;
-		_exponent >>= 1;
-	}
-	return result;
-}
-
-
-//
-// Init interpreter on entry.
-//
-void VM::initEntry()
-{
-	m_bounce = &VM::interpretCases;	
-	
-	interpretCases(); // first call initializes jump table
-	
-	initMetrics();
-	
-	copyCode();
-
-	optimize();
-}
-
-int VM::poolConstant(const u256& _con)
-{
-	TRACE_VAL(2, "pool constant", _con);
-	int i = 0, n = m_pool.size();
-	for (; i < n; ++i)
-	{
-		const u256& pooled = m_pool[i];
-		TRACE_VAL(2, "pooled constant", pooled);
-		if (_con == pooled)
-			return i;
-	}
-	if (i <= 255 )
-	{
-		TRACE_VAL(1, "constant pooled", _con);
-		m_pool.push_back(_con);
-		return i;
-	}
-	return -1;
-}
-
 void VM::optimize()
 {
-	size_t nBytes = m_codeSpace.size();
+	size_t pc, nBytes = m_ext->code.size();
 
 	// build a table of jump destinations for use in verifyJumpDest
 	
 	TRACE_STR(1, "Build JUMPDEST table")
-	for (size_t i = 0; i < nBytes; ++i)
+	for (pc = 0; pc < nBytes; ++pc)
 	{
-		Instruction op = Instruction(m_code[i]);
-		TRACE_OP(2, i, op);
+		Instruction op = Instruction(m_ext->code[pc]);
+		TRACE_OP(2, pc, op);
 				
 		// make synthetic ops in user code trigger invalid instruction if run
-		if (op == Instruction::PUSHC ||
-		    op == Instruction::JUMPV ||
-		    op == Instruction::JUMPVI)
+		if (
+			op == Instruction::PUSHC ||
+			op == Instruction::JUMPC ||
+			op == Instruction::JUMPCI
+		)
 		{
-			TRACE_OP(1, i, op);
-			m_code[i] = (byte)Instruction::BAD;
+			TRACE_OP(1, pc, op);
+			m_ext->code[pc] = (byte)Instruction::BAD;
 		}
 
 		if (op == Instruction::JUMPDEST)
 		{
-			m_jumpDests.push_back(i);
+			m_jumpDests.push_back(pc);
 		}
-		else if ((byte)Instruction::PUSH1 <= (byte)op &&
-		         (byte)op <= (byte)Instruction::PUSH32)
+		else if (
+			(byte)Instruction::PUSH1 <= (byte)op &&
+			(byte)op <= (byte)Instruction::PUSH32
+		)
 		{
-			i += (byte)op - (byte)Instruction::PUSH1 + 1;
+			pc += (byte)op - (byte)Instruction::PUSH1 + 1;
 		}
-		
+#ifdef EVM_JUMPS_AND_SUBS
+		else if (
+			op == Instruction::JUMPTO ||
+			op == Instruction::JUMPIF ||
+			op == Instruction::JUMPSUB)
+		{
+			++pc;
+			pc += 4;
+		}
+		else if (op == Instruction::JUMPV || op == Instruction::JUMPSUBV)
+		{
+			++pc;
+			pc += 4 * m_ext->code[pc];  // number of 4-byte dests followed by table
+		}
+		else if (op == Instruction::BEGINSUB)
+		{
+			m_beginSubs.push_back(pc);
+		}
+		else if (op == Instruction::BEGINDATA)
+		{
+			break;
+		}
+#endif
 	}
+	
+	copyCode(pc - nBytes);
```

Comment:
Actually, a PUSH code without enough data would have the same problem.  And worse - the copyCode() doesn't copy the data section at the end - what was I thinking?  But it can't, because it might all be zeros and all be gone.  So BEGINDATA needs a length argument.  And in the meantime it's back to nBytes+33.